### DISCUSSION TITLE
Transition plan for "dagger classic"

### DISCUSSION THREAD

**Discussion by shykes** [dagger/dagger#3432] [Ideas]
## Problem

We are developing Dagger SDKs for several languages, and rewriting the Dagger CLI to be a language-agnostic client, rather than a CUE-specific engine.

We intend to continue maintaining the original CLI (codenamed "dagger classic"), but we have not yet decided what to rename it to, and how to document and package it going forward.

## Solution

I propose the following transition plan for dagger classic:

* Introduce the concept of *Dagger CUE SDK*: everything required to develop for Dagger in CUE
* Rename the classic CLI to `?` (see "choice of new CLI name" section) and make it part of the CUE SDK
* Move the classic documentation in a new "CUE SDK" section, alongside the new "Go SDK" section. Edit the docs so that the wording matches
* Move the classic CLI code into the `main` branch, under the path `sdk/cue` (bikeshedding welcome)
* Rebase ongoing classic development work (most notably porting it to new engine by @jlongtine) to the `sdk/cue` directory
* Collaborate with upstream CUE developers to make sure the SDK follows CUE best practices. The more involved they are, the better.

## Choice of new CLI name

We have consensus on the following name: `dagger-cue`.

~~We must choose a new name for the classic CLI. Proposed options so far:~~

~~1. `cue-pipeline`~~
~~2. `cue-dagger`~~
~~3. `dagger-cue`~~

~~Please indicate your preference in the comments.~~

## Resources

A first draft of what docs for a CUE SDK might look like: https://github.com/dagger/dagger/pull/3431 (NOTE only structure changes, docs contents have not yet been edited to match the wording).

Feedback welcome!

**Comment by helderco** [comment]
Love the `sdk/cue` but I have one question.

I see two paths for CUE: 
1. backwards compatibility where your Europa plans still work on Cloak (hopefully, or with little change); 
2. new CUE that works more in tune with the current GraphQL API (like the other sdks are doing with codegen).

Have we decided if we're going to take both paths or just the first?

**Reply by jlongtine** [reply]
@helderco My read from this is both paths, FWIW.

**Reply by helderco** [reply]
Concurrently or as a later version?

**Reply by shykes** [reply]
Part 2 will be a later version. How much later depends on interest from the cue community, and involvement of the cue upstream devs. Purely because of our limited engineering resources.

**Reply by shykes** [reply]
Part 1 is a priority so that our current Cue developers can continue upgrading to the latest version of the engine.

**Reply by helderco** [reply]
Yep, I didn't mean concurrently now, just if both would exist at the same time at some point or if it would be supplanted, but you're right... depends on interest from the community. 🙂

**Comment by mircubed** [comment]
There are a few users that I can think of who'd like a CUE SDK. One example is here- https://github.com/dagger/dagger/issues/3121#issuecomment-1273080540

**Comment by shykes** [comment]
Any preferences between `cue-pipeline` and `cue-dagger`?

**Reply by jlongtine** [reply]
`cue-dagger`, personally.

**Reply by gerhard** [reply]
I'm thinking `dagger-cue`, since `dagger` is the top-level in this context.

**Reply by jlongtine** [reply]
I like @gerhard's suggestion.

**Reply by helderco** [reply]
I’m using `dagger-python` so it’s consistent with `dagger-cue`.

**Reply by shykes** [reply]
Looks like we have consensus on `dagger-cue`. Let me hunt for feedback to be extra sure, before confirming a decision.

**Comment by from-nibly** [comment]
I don't know how much I can contribute but if there's anything I can do especially testing out new versions I would like to help as much as I can. I have a docker builder repo built on classic dagger. I'm definitely ok with reworking it to get on the latest dagger cue sdk. But I would like to figure out how to help this moved so I don't feel stranded while waiting.
